STAC-22870: add reference docs and guide for otel component and relation mappings (#159)
rancher/stackstate-product-docs: docs/latest/modules/en/pages/setup/otel/otelmappings/schemas-ref.adoc

= Schemas
:revdate: 2025-12-23
:page-revdate: {revdate}
:description: Open Telemetry Mapping Schemas

== Overview

This page describes the schemas for defining an `OtelComponentMapping` or `OtelRelationMapping`, along with detailed explanations of constructs, expression syntax and semantics.

== Schemas for OTel Component & Relation Mappings

=== OtelComponentMapping

Each component mapping:

* Selects telemetry using conditions
* Extracts values using expressions

> * Produces a single logical component identified by a stable identifier

Reviewer: ```suggestion
* Selects telemetry using conditions.
* Extracts values using expressions.
* Produces a single logical component identified by a stable identifier.
```

  domainName: <cel-string>
  domainIdentifier: <cel-string>    # Optional
  layerName: <cel-string>
  layerIdentifier: <cel-string>     # Optional
  required:                         # Optional (required means that if any expression fails, the mapping fails)
    version: <cel-string>           # Optional
    additionalIdentifiers:          # Optional
      - <cel-string>
    tags:                           # Optional
      - source: <cel-string>
        target: string
      - source: <cel-string>
        pattern: regex
        target: string
  optional:                         # Optional (besides being optional on the schema, optional means that if any expression under `optional` fails, the mapping will continue, but without the failed expression/field)
    version: <cel-string>           # Optional
    additionalIdentifiers:          # Optional
      - <cel-string>
    tags:                           # Optional
      - source: <cel-string>
        target: string
      - source: <cel-map>
        pattern: regex
        target: string
expireAfter: duration-ms
----

=== OtelRelationMapping

Each relation mapping:

* Resolves a sourceId and targetId
* Assigns a relation type

> * Produces a directed edge between existing or future components

Reviewer: ```suggestion
* Resolves a `sourceId` and `targetId`.
* Assigns a relation type.
* Produces a directed edge between existing or future components.
```

    scope:
      condition: <cel-boolean>
      span | metric:
        condition: <cel-boolean>
        datapoint:
          condition: <cel-boolean>
          action: CREATE
vars:                               # Optional
  - name: string
    value: <cel-expression>
output:
  sourceId: <cel-string>
  targetId: <cel-string>
  typeName: <cel-string> 
  typeIdentifier: <cel-string>      # Optional
expireAfter: duration-ms
----

== Component & relation identity, merging, and lifecycle

=== Component identity

The `output.identifier` field defines the sole identity of a component. All components produced with the same identifier are considered the same logical entity and are merged by the platform.

This makes identifier construction a critical design choice. Identifiers should:

* Be stable across time
* Reflect the intended cardinality (service, instance, database, etc.)
* Avoid unbounded dimensions

[NOTE]
====
Using the `output.required.additionalIdentifiers` list, you can specify additional identifiers to associate the generated component with. This could be useful to correlate a component with an external identifier in another system.
====

=== Relation identity

> The relation identity is a composite of the form `output.sourceId-output.targetId`, where `sourceId` and `targetId` are component identifiers (e.g., from `output.identifier`).

Reviewer: ```suggestion
The relation identity is a composite of the form `output.sourceId-output.targetId`, where `sourceId` and `targetId` are component identifiers (for example, from `output.identifier`).
```

=== Refresh and expiration

Each mapping defines an expireAfter duration. This value controls how long a component or relation remains valid without being refreshed by new matching telemetry.

Internally:

> * The OTel collector continuously refreshes components as telemetry matches - this refresh window is based on the `expireAfter` field per mapping

Reviewer: ```suggestion
* The OTel collector continuously refreshes components as telemetry matches - this refresh window is based on the `expireAfter` field per mapping.
```

> * In the platform, components and relations expire if they are not refreshed in time (TODO: not sure what expiration config is used here - or if it's worth mentioning?)

Reviewer: [Placeholder reminder] - To remove the TODO mention before merging if there is nothing to add.

Reviewer 2: The platform expiration is determined by the `expireAfter` field. There is no additional expiration config.  So this bullet can be removed.

== Input traversal model

Topology mappings traverse OpenTelemetry data hierarchically:

> `resource → scope → metric/span → datapoint`

Reviewer: ```suggestion
Topology mappings traverse OpenTelemetry data hierarchically - `resource → scope → metric/span → datapoint`
```

* datapoint.attributes

Examples:

[source]
----
metric.name == 'traces_service_graph_request_total'
# OR
datapoint.attributes['connection_type'] == 'database'
----

**LOGS**

Not yet supported for topology mappings.

=== Conditions and actions

**Conditions**

Conditions determine whether processing continues at a given level. If a condition evaluates to false, the mapping is skipped for that telemetry element.

Conditions may only reference fields available at that level. For example, the following snippet for the input block would be invalid:

[,yaml]
----
input:
  signal:
    - "TRACES"
  resource:
    condition: "${'service.name' in resource.attributes}" # action omitted since the default is CONTINUE
    scope:
      condition: "${resource.attributes['service.name'] == 'cart-svc'}" # it's not allowed to access resource-level fields at scope-level
      action: "CREATE" # input block describes that it expects to filter until this level only
----

**Actions**

Supported actions:

* CONTINUE – continue evaluation to the next level
* CREATE – create a component or relation at this level

Defaults:

* Conditions default to true
* Actions default to CONTINUE

> * At terminal levels (span, datapoint), the default action is CREATE

Reviewer: ```suggestion
* Conditions default to 'true'.
* Actions default to 'CONTINUE'.
* At terminal levels (span, datapoint), the default action is 'CREATE'.
```

This means a component or relation is only produced when the terminal condition evaluates to true.

Constraints:

* When there are multiple input signals declared, data (using expression) can only be accessed from the common-ancestor-level, which in all cases are scope-level.
* Data access from another input-signal is not allowed. For example, metric-level fields cannot be accessed from span-level.

TODO: more input examples?

== Variables (vars)

Mappings may define variables to avoid repetition and improve readability.

Variables:

* Are evaluated using CEL expressions
* May reference any fields available at the current level

> * Can be reused across output fields

Reviewer: ```suggestion
* Are evaluated using CEL expressions.
* May reference any fields available at the current level.
* Can be reused across output fields.
```

  - name: string
    value: <cel-expression>
output:
  sourceId: <cel-string>
  targetId: <cel-string>
  typeName: <cel-string> 
  typeIdentifier: <cel-string>      # Optional
expireAfter: duration-ms
----

== Component & relation identity, merging, and lifecycle

=== Component identity

The `output.identifier` field defines the sole identity of a component. All components produced with the same identifier are considered the same logical entity and are merged by the platform. This makes identifier construction a critical design choice. 

Identifiers should:

* Be stable across time.
* Reflect the intended cardinality (service, instance, database, and the like).
* Avoid unbounded dimensions.

[NOTE]
====
Using the `output.required.additionalIdentifiers` list, you can specify additional identifiers to associate the generated component with. This could be useful to correlate a component with an external identifier in another system.
====

=== Relation identity

The relation identity is a composite of the form `output.sourceId-output.targetId`, where `sourceId` and `targetId` are component identifiers (for example, from `output.identifier`).

=== Refresh and expiration

> Each mapping defines an `expireAfter` duration. This value controls how long a component or relation remains valid without being refreshed by new matching telemetry.

Reviewer 2: This does not just control the refreshing, it means that the component is removed from the topology when it expires.

```suggestion
Each mapping defines an `expireAfter` duration. This value controls how long a component or relation remains in the topology without being refreshed by new matching telemetry. If no refresh happens the component or relation is removed from the topology.
```

**TRACES**

Provides access to:

* resource.attributes
* scope.name, scope.version, scope.attributes
* span.name, span.kind, span.statusMessage, span.statusCode, span.attributes

Examples:

[source]
----
resource.attributes['service.name'] == 'checkout'
# OR
span.kind == 'SPAN_KIND_SERVER' && span.attributes['http.method'] == 'POST'
----

**METRICS**

Provides access to:

* resource.attributes
* scope.name, scope.version, scope.attributes
* metric.name, metric.description, metric.unit
* datapoint.attributes

Examples:

[source]
----
metric.name == 'traces_service_graph_request_total'
# OR
datapoint.attributes['connection_type'] == 'database'
----

=== Conditions and actions

**Conditions**

Conditions determine whether processing continues at a given level. If a condition evaluates to false, the mapping is skipped for that telemetry element.

Conditions may only reference fields available at that level. For example, the following snippet for the input block would be invalid:

[,yaml]
----
input:
  signal:
    - "TRACES"
  resource:
    condition: "${'service.name' in resource.attributes}" # action omitted since the default is `CONTINUE`
    scope:

>       condition: "${resource.attributes['service.name'] == 'cart-svc'}" # it's not allowed to access resource-level fields at scope-level

Author: Yeah, you're right. Fixed.

      action: "CREATE" # input block describes that it expects to filter until this level only
----

**Actions**

Supported actions:

* CONTINUE – continue evaluation to the next level.
* CREATE – create a component or relation at this level.

Defaults:

* Conditions default to 'true'.
* Actions default to 'CONTINUE'.

> * At terminal levels (span, datapoint), the default action is 'CREATE'.

Reviewer 2: I think a `resource` , `metric` or `scope` can also be a terminal level if there is nothing nested deeper. I would expect those to also act like a `CREATE` in those cases.

Though now that I think about it, I thought we discarded the whole idea of having a default CREATE action. Instead preferring it to be explicit to avoid automatically traversing all data.

But if the behavior is that we can omit the CREATE action at `span` or `datapoint` I guess that's ok although a bit surprising. As long as it only works when those levels are actually specified in the yaml.

Author: Good catch. The wording is not correct. Technically, if the action is omitted, the implementation defaults to continue - which results in the mapping being skipped as there's no explicit CREATE specified.

Here are a few more examples of valid input declarations:

**TRACES**

[,yaml]
----
input:
  signal:
    - "TRACES"
  resource:
    condition: "${'service.name' in resource.attributes}" # action omitted since the default is `CONTINUE`
    scope: # action and condition have been omitted since the defaults are `CONTINUE` and `true` respectively
      span:
        action: "CREATE"
        condition: "${span.attributes['http.method'] == 'POST'}"
----

**METRICS**

[,yaml]
----
input:
  signal:
    - "METRICS"
  resource: # action and condition have been omitted since the defaults are `CONTINUE` and `true` respectively
    scope:
      condition: "scope.name == 'traces_service_graph'"
      metric: # action omittted since the default is `CONTINUE`
        condition: "metric.name == 'traces_service_graph_request_total'"
        datapoint:
          action: "CREATE"
          condition: |
            'client' in datapoint.attributes &&
            'server' in datapoint.attributes
----

== Variables (vars)

Mappings may define variables to avoid repetition and improve readability.

Variables:

* Are evaluated using CEL expressions.

> * May reference any fields available at the current level.

Reviewer 2: "Current level" is a bit unclear, the variables are defined outside of the nested levels. I was going to suggest that variables can use the same fields as output fields can, but I can't find that documented anywhere.

It actually looks like there is nothing at all on the `output` section, the docs dive into `tag mappings` directly. Maybe include a short section on that too and then for both output and vars sections we can say that they "can use fields that are available at the deepest level matched by the input filters, for example if the input selection has the Create action at the span level, resource, scope and span fields can be used".

Author: Good catch.
